torch/_inductor/codegen/cpp.py

```diff
@@ -1662,42 +1727,65 @@ def gen_kernel(kernel):
+                            return suffix
                         else:
-                            return kernel.reduction_prefix
-                return None
+                            prefix = kernel.reduction_prefix
+                            if loop.parallel:
+                                prefix = prefix + kernel.parallel_reduction_prefix
+                            return prefix
 
             def gen_loops(loops: List[LoopLevel], in_reduction=False):
                 with contextlib.ExitStack() as stack_outer:
+                    local_reduction_init, local_reduction_stores = None, None
                     if loops:
                         loop = loops[0]
-                        if loop.is_reduction() and not in_reduction:
-                            reduction_prefix = get_reduction_code_buffer(
-                                loops, is_suffix=False
-                            )
+                        if loop.is_reduction and not in_reduction:
+                            (
+                                local_reduction_init,
+                                local_reduction_stores,
+                            ) = get_reduction_code_buffer(loops, "local")
+                            reduction_prefix = get_reduction_code_buffer(loops)
+                            if local_reduction_init:
+                                assert local_reduction_stores
                             if reduction_prefix:
                                 stack_outer.enter_context(code.indent())
                             code.splice(reduction_prefix)
                         if loop_nest.is_reduction_only() and loop.parallel:
                             worksharing.parallel(threads)
+                            if local_reduction_init:
+                                code.splice(local_reduction_init)
 
                     for loop in loops:
-                        gen_loop(loop, in_reduction)
+                        gen_loop(loop)
 
                     if loops:
                         loop = loops[0]
                         if loop_nest.is_reduction_only() and loop.parallel:
+                            if local_reduction_stores:
+                                code.splice(local_reduction_stores)
                             worksharing.close()
-                        if loop.is_reduction() and not in_reduction:
-                            code.splice(
-                                get_reduction_code_buffer(loops, is_suffix=True)
-                            )
+                        if loop.is_reduction and not in_reduction:
+                            code.splice(get_reduction_code_buffer(loops, "suffix"))
+
+            def gen_loop(loop: LoopLevel):
+                def is_parallel_reduction(loop):
+                    while loop.parent:
+                        loop = loop.parent
+                    return loop.is_reduction and loop.parallel
```

Comment:
is checking `loop.is_reduction and loop.parallel` good enough here?

Reply:
I met to case not able to work here
```cpp
                #pragma omp for
                for(long x0=static_cast<long>(0L); x0<static_cast<long>(48L); x0+=static_cast<long>(16L))
                {
                    for(long x1=static_cast<long>(0L); x1<static_cast<long>(48L); x1+=static_cast<long>(16L))
                    {
                        float tmp0[16*16] __attribute__ ((aligned (16)));
                        at::vec::transpose_mxn<float,16,16>(in_ptr0 + static_cast<long>(x1 + (50L*x0)), static_cast<long>(50L), tmp0, 16);
                        for (long x1_inner = 0; x1_inner < 16; x1_inner++) // loop is reduction but not parallel
                        {
                            auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + static_cast<long>(16L*x1_inner));
                            auto tmp2 = at::vec::Vectorized<float>::loadu(in_ptr1 + static_cast<long>(x0 + (50L*x1) + (50L*x1_inner)));
                            auto tmp3 = tmp1 * tmp2;
                            tmp_acc0_vec_local = tmp_acc0_vec_local + tmp3;
                        }

```

Reply:
Thanks. Make sense. Can you create a `get_root()` method in the `LoopLevel` so that we can simplify the logic here as below?
```python
root = loop.get_root()
return root.is_reduction and root.parallel
```